scheduler/src/scheduling.test.ts

```diff
@@ -1,111 +1,105 @@
 // Type assertion is ok here, because we have mocked the whole './clients/adapter-client' module
 /* eslint-disable @typescript-eslint/consistent-type-assertions */
-const mockedGetLatestEventId = getLatestEventId as jest.Mock
-mockedGetLatestEventId.mockResolvedValue(321)
 const mockedGetAllDatasources = getAllDatasources as jest.Mock
-const mockedGetEventsAfter = getEventsAfter as jest.Mock
-const mockedGetDatasource = getDatasource as jest.Mock
+const mockedTriggerDatasource = triggerDatasource as jest.Mock
 /* eslint-enable @typescript-eslint/consistent-type-assertions */
 
+const CONNECTION_RETRIES = 2
+const CONNECTION_BACKOFF_IN_MS = 1000
+const TRIGGER_RETRIES = 3
+
 jest.mock('./env', () => () => ({
   MAX_TRIGGER_RETRIES: 2
 }))
+mockedTriggerDatasource.mockReturnValue(Promise.resolve())
+
+let scheduler: Scheduler
 
 describe('Scheduler', () => {
+  beforeEach(async () => {
+    scheduler = new Scheduler(TRIGGER_RETRIES)
+  })
+
+  afterEach(async () => {
+    mockedTriggerDatasource.mockClear()
+  })
+
   test('should initialize jobs correctly', async () => {
-    const config = generateConfig(true, new Date(Date.now() + 5000), 6000)
+    const config = generateConfig(true, new Date(Date.now() + 500), 6000)
     mockedGetAllDatasources.mockResolvedValue([config])
+    await scheduler.initializeJobsWithRetry(CONNECTION_RETRIES, CONNECTION_BACKOFF_IN_MS)
 
-    await Scheduling.initializeJobsWithRetry()
+    await sleep(1000)
 
-    expect(Scheduling.getAllJobs()).toHaveLength(1)
-    expect(Scheduling.getAllJobs()[0].datasourceConfig).toEqual(config)
+    expect(scheduler.getAllJobs()).toHaveLength(1)
+    expect(scheduler.getAllJobs()[0].datasourceConfig).toEqual(config)
+    expect(mockedGetAllDatasources).toHaveBeenCalledTimes(1)
+    expect(mockedTriggerDatasource).toHaveBeenCalledTimes(1)
   })
 
   test('should apply creation event', async () => {
     mockedGetAllDatasources.mockResolvedValue([])
-    const toBeAdded = generateConfig(true, new Date(Date.now() + 5000), 6000)
-    const creationEvent = {
-      eventId: 322,
-      eventType: EventType.DATASOURCE_CREATE,
-      datasourceId: 123
+    const toBeAdded = generateConfig(true, new Date(Date.now() + 500), 6000)
+    const creationEvent: DatasourceConfigEvent = {
+      datasource: toBeAdded
     }
 
-    await Scheduling.initializeJobsWithRetry()
-    expect(Scheduling.getAllJobs()).toHaveLength(0)
+    await scheduler.applyCreateOrUpdateEvent(creationEvent)
 
-    mockedGetEventsAfter.mockResolvedValue([creationEvent])
-    mockedGetDatasource.mockResolvedValue(toBeAdded)
-    await Scheduling.updateDatasources()
+    await sleep(500)
 
-    await sleep(1000) // Give the scheduler some time to apply the update
-
-    expect(Scheduling.getAllJobs()).toHaveLength(1)
-    const job123 = Scheduling.getJob(123)
+    expect(scheduler.getAllJobs()).toHaveLength(1)
+    const job123 = scheduler.getJob(123)
     expect(job123).toBeDefined()
     if (job123 !== undefined) {
```

Comment:
A if statement which exclude a `expect()` statement should not be present in a test.

`expect(job123).toBeDefined()` checks already if the job is not `undefined`. 

I would remove the if statement and just write the check inside the `{}` direct after the other.

Reply:
I think the if-clause was added because the typescript compiler does not recognize the non-null check by `expect()`. I will suppress the ts error by `@ts-suppress.`